tests/test_typed_struct.py

```diff
@@ -0,0 +1,439 @@
+    class Foo(ts.TypedStruct):
+        a = int
+        b = int
+
+    class Bar(ts.TypedStruct):
+        a = int
+        b = int
+
+    assert Foo(a=1, b=2) == Foo(a=1, b=2)
+    assert Foo(a=1, b=2) != Foo(a=1, b=3)
+    assert Foo(a=1, b=2) != Bar(a=1, b=2)
+
+    assert Foo(a=1, b=2).to_dict() == Bar(a=1, b=2).to_dict()
+    assert Foo(a=1, b=2).to_bunch() == Bar(a=1, b=2).to_bunch()
+
+
+@pytest.mark.parametrize("use_full_syntax", [True, False])
+def test_typed_struct_list_fields(use_full_syntax):
+    class Foo(ts.TypedStruct):
+        a = int
+
+    if use_full_syntax:
+        class Bar(ts.TypedStruct):
+            foos = ts.Field([Foo])
+            nums = ts.Field([int])
+    else:
+        class Bar(ts.TypedStruct):
+            foos = [Foo]
+            nums = [int]
+
+    bar = Bar()
+    assert bar.foos == []
+
+    # append()
+    bar.foos.append(Foo(a=1))
+    bar.foos.append(dict(a=2))
+    with pytest.raises(ts.FieldTypeMismatch):
+        bar.foos.append(3)
+    assert bar.foos == [Foo(a=1), Foo(a=2)]
+
+    bar.nums.append(1)
+    with pytest.raises(ts.FieldTypeMismatch):
+        bar.nums.append('2')
+    assert bar.nums == [1]
+
+    # __setitem__()
+    bar.foos[0] = Foo(a=4)
+    bar.foos[1] = dict(a=5)
+    with pytest.raises(ts.FieldTypeMismatch):
+        bar.foos[0] = 6
+    assert bar.foos == [Foo(a=4), Foo(a=5)]
+
+    bar.nums[0] = 3
+    assert bar.nums == [3]
+
+    # insert()
+    bar.foos.insert(0, Foo(a=7))
+    bar.foos.insert(2, dict(a=8))
+    with pytest.raises(ts.FieldTypeMismatch):
+        bar.foos.insert(1, 9)
+    assert bar.foos == [Foo(a=7), Foo(a=4), Foo(a=8), Foo(a=5)]
+
+    bar.nums.insert(0, 4)
+    assert bar.nums == [4, 3]
+
+    # Assignment
+    bar.foos = [Foo(a=10), dict(a=11)]
+    with pytest.raises(ts.FieldTypeMismatch):
+        bar.foos = [12]
+    assert bar.foos == [Foo(a=10), Foo(a=11)]
+
+    bar.nums = [5]
+    assert bar.nums == [5]
+
+    # __iadd__()
+    bar.foos += [Foo(a=13), dict(a=14)]
+    with pytest.raises(ts.FieldTypeMismatch):
+        bar.foos += [15]
+    assert bar.foos == [Foo(a=10), Foo(a=11), Foo(a=13), Foo(a=14)]
+
+    bar.nums += [6]
+    assert bar.nums == [5, 6]
+
+    # extend()
+    bar.foos.extend(iter([Foo(a=16), dict(a=17)]))
+    with pytest.raises(ts.FieldTypeMismatch):
+        bar.foos.extend(range(18))
+    assert bar.foos == [Foo(a=10), Foo(a=11), Foo(a=13), Foo(a=14), Foo(a=16), Foo(a=17)]
+
+    bar.nums.extend(iter([7]))
+    assert bar.nums == [5, 6, 7]
+
+    # Make sure that bar.foos and bar.nums are still a ListCollection
+    assert bar.foos.M.a.C == [10, 11, 13, 14, 16, 17]
+    assert bar.nums.M.call(lambda x: x * 10).C == [50, 60, 70]
+
+
+@pytest.mark.parametrize("use_full_syntax", [True, False])
+def test_typed_struct_dict_fields(use_full_syntax):
+    class Foo(ts.TypedStruct):
+        a = int
+
+    if use_full_syntax:
+        class Bar(ts.TypedStruct):
+            foos = ts.Field({int: Foo})
```

Comment:
just so I'm clear - if we could we'd do `ts.Field(Dict[int, Foo])`, right?

Reply:
Right. And `ts.Field(Bunch[Foo])` instead of `ts.Field({str: Foo})` too.